src/mlang/m_ir/mir.ml

```diff
@@ -287,6 +306,7 @@ type exec_pass = { exec_pass_set_variables : literal Pos.marked VariableMap.t }
 
 type program = {
   program_vars : variable_data VariableMap.t;
+  program_rules : rule_data RuleMap.t;
```

Comment:
I would prefer this type to reflect more the structure of the programs with only one rule map where each value contains a little map of the variables defined inside this rule.

Reply:
In fact, inside a given rule, we should store a list of variable assignations and not a map, since the order is important.